Adding blog for date histogram optimizations (#3782)
opensearch-project/project-website: _posts/2025-05-05-date-histogram-optimizations.md

![Aggregation execution flow](../assets/media/blog-images/2025-05-05-date-histogram-optimizations/Aggregation_Execution_Flow.png){:class="img-centered" width="650px"}

By default, OpenSearch processes aggregations by first evaluating each query condition on every shard using Lucene, which builds iterators that identify matching document IDs. These iterators are then combined (e.g., using logical AND) to find documents that satisfy all query filters. The resulting set of matching document IDs is streamed to the aggregation framework, where each document flows through one or more aggregators. These aggregators use Lucene’s doc values (explained above) to efficiently retrieve the field values needed for computation (e.g., for calculating averages or counts). This streaming model is hierarchical—documents pass through a pipeline of aggregators, allowing them to be grouped into top-level and nested buckets simultaneously. For example, a document can first be bucketed by month, then further aggregated by HTTP status code within that month. This design enables OpenSearch to process complex, multi-level aggregations efficiently in a single pass over the matching documents.

> ## Understanding the setup

Reviewer: I feel it maybe better to organize 
`How numeric data is stored in OpenSearch`
`How Aggregation works in OpenSearch`
under this section

```json
{
  "size": 0,
  "query": {
    "range": {
      "dropoff_datetime": {
        "gte": "2015-01-01 00:00:00",
        "lt": "2016-01-01 00:00:00"
      }
    }
  },
  "aggs": {
    "dropoffs_over_time": {
      "date_histogram": {
        "field": "dropoff_datetime",
        "calendar_interval": "month"
      }
    }
  }
}
```
While the number of nodes does not matter, we worked with single node cluster containing all the data of nyc_taxis dataset across 1 shard without any replicas. 

## Performance Bottlenecks

When we began analyzing performance, we started single date histogram aggregation query execution in loop and collected flame graphs during the execution [9310](https://github.com/opensearch-project/OpenSearch/issues/9310). The flame graph primarily indicated towards two key limitations:
1. **Data Volume Dependency**: Query latency was directly proportional to data volume. For example, a one-month aggregation taking 1 second would take 12 seconds for a year's worth of data.

> 2. **Bucket Count Impact**: Large numbers of buckets (like in minute-level aggregations) led to hash collision issues, further degrading performance.

Reviewer: Not aware of hash collision issues before, did we know this is the issue?

### Phase 1: Filter Rewrite

Based on our understanding of range query execution, we came up with Filter Rewrite approach. It preemptively created a series of range filters, one for each bucket, in the requested date histogram. For example, the monthly aggregation over year can be rewritten as below:

```json
{
  "size": 0,
  "aggs": {
    "dropoffs_over_time ": {
      "filters": {
        "1420070400000": {
          "range": {
            "dropoff_datetime": {
              "gte": "2015-01-01 00:00:00",
              "lt": "2015-02-01 00:00:00"
            }
          }
        },
        "1422748800000": {
          "range": {
            "dropoff_datetime": {
              "gte": "2015-02-01 00:00:00",
              "lt": "2015-03-01 00:00:00"
            }
          }
        },
        "1425168000000": {
          "range": {
            "dropoff_datetime": {
              "gte": "2015-03-01 00:00:00",
              "lt": "2015-04-01 00:00:00"
            }
          }
        },
        ...
      }
    }
  }
}
```

> This date histogram aggregation query generates filter corresponding to each bucket and leverages Lucene’s Points Index, which uses BKD trees, to significantly optimize the aggregation. This tree-based structure organizes data into nodes representing value ranges with associated document counts, enabling efficient traversal. By skipping irrelevant subtrees and leveraging early termination, the system reduces unnecessary disk reads and avoids visiting individual documents. which are executed in similar manner as range query using the index tree to determine counts for each bucket faster than document value iteration. Similar approach was also applied to the auto date histogram, composite aggregation on date histogram source, and, later on, numeric range aggregation with consecutive ranges (TODO: Why do we need consecutive ranges??).

Reviewer: > Why do we need consecutive ranges
It's the outcome of a date histogram.

Author: does numeric range aggregation query also need that? IMO, if consecutive ranges is not the requirement of this approach, we need not make that classification

Reviewer: Range agg doesn't have such requirement - consecutive ranges. 
The multi-range traversal algorithm should still be able to support in-consecutive ranges case with some modification, like maintain a list of ranges that are active regarding the current travelling point.

Author: @bowenlan-amzn - Do you still have any concerns with this?

Reviewer: I believe we are on the same page. We don't need to mention `consecutive ranges` as you said.

![Performance_Improvements_Initial.png](../assets/media/blog-images/2025-05-05-date-histogram-optimizations/Performance_Improvements_Initial.png)

> The following diagrams illustrate how documents are counted per histogram bucket using the index tree. To efficiently count documents matching a range (e.g., 351 to 771), the traversal begins at the root, checking whether the target range intersects with the node’s range. If it does, the algorithm recursively explores the left and right subtrees. An important optimization is skipping entire subtrees: if a node’s range falls completely outside the query range (e.g., 1–200), it is ignored. Conversely, if a node’s range is fully contained within the query range (e.g., 401–600), the algorithm can return the document count from that node directly without traversing its children. This allows the engine to avoid visiting all leaf nodes, focusing only on nodes with partial overlaps. As a result, the operation becomes significantly faster—reducing complexity from linear (O(N)) to logarithmic (O(log N)) in many cases—since it leverages the hierarchical structure to skip large irrelevant portions of the tree and aggregate counts efficiently.

Reviewer: I don't feel (O(N)) to (O(log N)) is a good way to tell the difference. The real time complexity is much complicated than that. 
I think use the tree to quickly find the target range start, and skip sub trees are good and simple explanation already.

Author: Fair point, I have removed all the references to exact time complexity from this blog

![MRT_Initial.png](../assets/media/blog-images/2025-05-05-date-histogram-optimizations/MRT_Initial.png)
![MRT_Middle.png](../assets/media/blog-images/2025-05-05-date-histogram-optimizations/MRT_Middle.png)

> ![MRT_Final.png](../assets/media/blog-images/2025-05-05-date-histogram-optimizations/MRT_Final.png)

Reviewer: Seems there are no go-through explanation of these diagrams

Author: Yeah, @kolchfa-aws also added that feedback. Let me work on addressing that

---
layout: post
title:  "Optimizing date histogram aggregations using the index tree OR Transforming Bucket Aggregations: Our Journey to 100x Performance Improvement"
authors:
   - akjain
   - bowenlan
   - kolchfa

> date: 2025-05-05

Reviewer 2: Blog publish date: 2025-07-24

            "dropoff_datetime": {
              "gte": "2015-02-01 00:00:00",
              "lt": "2015-03-01 00:00:00"
            }
          }
        },
        "1425168000000": {
          "range": {
            "dropoff_datetime": {
              "gte": "2015-03-01 00:00:00",
              "lt": "2015-04-01 00:00:00"
            }
          }
        },
        ...
      }
    }
  }
}
```

This date histogram aggregation query generates a filter corresponding to each bucket and uses Lucene's Points Index, which is based on BKD trees, to significantly optimize the aggregation. This tree-based structure organizes data into nodes representing value ranges with associated document counts, enabling efficient traversal. By skipping irrelevant subtrees and using early termination, the system reduces unnecessary disk reads and avoids visiting individual documents. The counts for each bucket are determined using the index tree, similarly to a `range` query, which is faster than iterating over document values. We also applied this approach to the auto-date histogram, composite aggregation on date histogram source, and, later on, numeric range aggregation with non-overlapping ranges.

> ![Performance_Improvements_Initial.png](/assets/media/blog-images/2025-05-05-date-histogram-optimizations/Performance_Improvements_Initial.png)

Author: Good catch! I have added text referencing this graph, kindly review the same

* While multi-range traversal significantly reduced overhead, extremely fine-grained histograms over sparse datasets can potentially still encounter performance regressions.

## Looking forward

These optimizations have been contributed back to Lucene ([this pull request](https://github.com/apache/lucene/pull/14439)), so that other search systems like Elasticsearch and Solr can benefit from these improvements. We are continuing to work on performance improvements to address the limitations in these areas:
* Nested aggregations
* Multi-field queries
* More efficient handling of deleted documents

## Conclusion

Index-based optimization for date histogram aggregations in OpenSearch significantly enhances the performance of time-series analysis and visualization. It is applied automatically to eligible aggregations, streamlining your workflows without extra manual effort. As OpenSearch evolves, these improvements ensure that you can efficiently gain insights from your data, with less concern about scale.

> This journey shows how iterative improvements, deep system understanding, and community collaboration can lead to breakthrough performance gains. While we initially didn't expect such dramatic results, our commitment to continuous optimization paid off in ways we couldn't have imagined.

Author: I have restructured slightly and moved `Looking Forward` at the bottom. Please let me know what you think

}
```

> This date histogram aggregation query generates a filter corresponding to each bucket and uses Lucene's Points Index, which is based on BKD trees, to significantly optimize the aggregation. This tree-based structure organizes data into nodes representing value ranges with associated document counts, enabling efficient traversal. By skipping irrelevant subtrees and using early termination, the system reduces unnecessary disk reads and avoids visiting individual documents. The counts for each bucket are determined using the index tree, similarly to a `range` query, which is faster than iterating over document values. We also applied this approach to the auto-date histogram, composite aggregation on date histogram source, and, later on, numeric range aggregation with non-overlapping ranges. The below graph shows the performance improvement for daily and hourly date histogram aggregations compared to OS 2.7 and OS 2.11. Also, notice that the minutely aggregation (gray bar in below graph) could not initially benefit from this optimization. We talk about that in more detail in the next section.

Reviewer 3: ```suggestion
This date histogram aggregation query generates a filter corresponding to each bucket and uses Lucene's Points Index, which is based on BKD trees, to significantly optimize the aggregation. This tree-based structure organizes data into nodes representing value ranges with associated document counts, enabling efficient traversal. By skipping irrelevant subtrees and using early termination, the system reduces unnecessary disk reads and avoids visiting individual documents. The counts for each bucket are determined using the index tree, similarly to a `range` query, which is faster than iterating over document values. We also applied this approach to the auto-date histogram, composite aggregation on date histogram source, and, later on, numeric range aggregation with non-overlapping ranges. The following graph shows the performance improvement for daily and hourly date histogram aggregations compared to OpenSearch 2.7 and 2.11. Also notice that the minute-level aggregation (gray bar in the graph) could not initially benefit from this optimization. We discuss this in more detail in the next section.
```

* While multi-range traversal significantly reduced overhead, extremely fine-grained histograms over sparse datasets can potentially still encounter performance regressions.

## Conclusion

Index-based optimization for date histogram aggregations in OpenSearch significantly enhances the performance of time-series analysis and visualization. It is applied automatically to eligible aggregations, streamlining your workflows without added manual effort. As OpenSearch evolves, these improvements ensure that you can efficiently gain insights from your data, with less concern about scale.

This journey shows how iterative improvements, deep system understanding, and community collaboration can lead to breakthrough performance gains. While we initially didn't expect such dramatic results, our commitment to continuous optimization paid off in ways we couldn't have imagined.

## Looking forward

> These optimizations have been contributed upstream to Lucene (see [this pull request](https://github.com/apache/lucene/pull/14439)) so that other search systems such as Elasticsearch and Solr can also benefit from these improvements. Looking ahead, we’re continuing to explore and implement further performance enhancements, particularly in the following areas:

Reviewer 3: ```suggestion
These optimizations have been contributed upstream to Lucene (see [this pull request](https://github.com/apache/lucene/pull/14439)) so that other search systems such as Elasticsearch and Solr can also benefit from these improvements. Looking ahead, we're continuing to explore and implement further performance enhancements, particularly in the following areas:
```

* Nested aggregations
* Multi-field queries
* More efficient handling of deleted documents

> If you're interested in collaborating or discussing these topics further, feel free to reach out via GitHub or Slack—we’d love to connect.

Reviewer 3: ```suggestion
If you're interested in collaborating or discussing these topics further, feel free to reach out on GitHub or Slack—we'd love to connect!
```
